### DISCUSSION TITLE
[Cross-Post] MAISI ControlNet Semantic Mask Conditioning Training Issues

### DISCUSSION THREAD

**Discussion by pshavela** [Project-MONAI/tutorials#1991] [Q&A]
Hello all,
I will cross-post this from the main repository ( [#8451](https://github.com/Project-MONAI/MONAI/discussions/8451#discussion-8316146) )
I am currently working on my master's project and it is a bit urgent:

firstly I would like to thank the MAISI team for their amazing work.
I have been playing around with it and I am currently trying to reproduce the mask-conditioned synthesis on a coronary CT dataset. I have trained the VAE and the MAISI diffusion model from scratch according to the tutorials, with great success.

Unfortunately, I cannot get the ControlNet to converge, I have tried different learning rates (1e-5 and 1e-6), let it train
for more than 20000 steps, however, as the images below show the conditioning is not applied well enough.
I am also comparing the conditionally generated images at every epoch, but they yield similar results.

Does someone have some pointers on how to properly train the ControlNet with semantic masks and how to achieve comparable results?

![443087953-adeb881d-6037-44e2-b0e8-3f2ae6316778](https://github.com/user-attachments/assets/ccef20b4-fd3b-470d-9d0b-a3ae698f2181)
![443088050-d63ff891-984f-4a13-b34c-bb5f1d6d03e3](https://github.com/user-attachments/assets/d89112bc-4f59-4e70-8805-d5da1926d4b2)
![443088053-64a39b4b-6589-461e-928e-3ac7859ab7db](https://github.com/user-attachments/assets/3eb444e1-aeeb-40c3-b979-1dc2723271c5)

**Comment by Can-Zhao** [comment]
Did you see it get better after training it longer?

**Reply by pshavela** [reply]
Hello, thank you for reaching out!
Yes, I have gone up to 750,000 steps, but unfortunately it produced similar loss patterns and inference with the semantic mask was similar to the one above.

**Reply by Can-Zhao** [reply]
May I ask if you have binarized the mask as the controlnet training script shows? Thank you!

**Reply by sara-create** [reply]
Hello @Can-Zhao ! I have the same problem , my loss doesn't decrease even after a long time and I don't understand why . I binarize the masks and follow the script ( maisi_train_controlnet_tutorial.ipynb ) but it doesn't converge after a lot of trials. For the context , I have 299 CT-scans with their masks.

**Reply by sara-create** [reply]
Hello  @Can-Zhao ! I have the same problem , my loss doesn't decrease even after a long time and I don't understand why ( it oscillates a lot ). I binarize the masks and follow the script ( maisi_train_controlnet_tutorial.ipynb ) but it doesn't converge after a lot of trials. For the context , I have 299 CT-scans with their masks.

**Reply by sara-create** [reply]
@pshavela did you manage to solve the problem at the end ?

**Reply by pshavela** [reply]
@sara-create Hi, I would say the results were okay-ish, however, the same problems regarding loss values persisted. Our hypothesis is that our dataset of 700 images is simply too small, and the finetuning of the controlnet is simply not robust, thus the controlnet model is not able to fully extrapolate. 

We instead opted for a joint training of a hypernet and underlying diffusion 
model, both trained from scratch at once. The results were better in terms of semantic conditioning and preservation of annotated organs.

 But my thesis project has concluded, so I have not pursued this problem any further.

**Reply by Can-Zhao** [reply]
It's true that loss of DDPM does not mean much

**Reply by sara-create** [reply]
@pshavela Thank you so much for your answer ! I really appreciate it. So at the end the solution that worked best for you was the hypernet along with the DDPM is that right ? You retrained the DDPM from scratch ? Was it enough with 700 volumes ? I am using their checkpoint since I thought , in my case , that 300 volumes weren't enough to retrain the DDPM from scratch

**Reply by pshavela** [reply]
@sara-create Exactly, we trained the DDPM and the hypernet both from scratch at the same time. Keep in mind that
this requires more GPU memory than training seperately. And yes, 700 images for training was enough. 

But i would also like to mention that we switched to a flow matching-based approach (rectified flow) instead of DDPM, as it proved to be superior in terms of training (500 epochs was enough) and also inference speeds. we used image sizes of 256x256x128.

**Reply by sara-create** [reply]
Thank you so much @pshavela !! One more last question , have you retrained the VAE or used their checkpoint ?

**Reply by Can-Zhao** [reply]
One possible reason that fune-tuning controlnet did not work: The pretrained controlnet is based on a certain label definition [./configs/label_dict.json](https://github.com/Project-MONAI/tutorials/blob/main/generation/maisi/configs/label_dict.json)
If your data label did not follow it, then fine-tuning from MAISI controlnet does not make sense. Training controlnet from scratch is the way to go. But you can still reuse VAE and diffusion model.

If your data label already followed it, but still not working, then maybe trying rectified flow could be a direction.

**Reply by pshavela** [reply]
@sara-create no problem, yes, we did retrain the VAE 👍

**Comment by Mudassar-MLE** [comment]
Hi @Can-Zhao ,

First, thank you and the MAISI team for the incredible work. I am currently exploring MAISI for neck CT, but I’m running into several challenges. There don’t seem to be many anatomies supported for head/neck CT. Even the existing models are unable to generate meaningful images of the brain, head, or neck although the documentation mentions that indexing “1” should support head-neck CT.

I tried fine-tuning ControlNet following the provided instructions, adding around 40 new labels with 5–10 cases. However, the model did not converge. After experimenting with different parameters and epochs, it showed slight convergence, but the outputs still mostly resembled chest anatomy with a rough neck contour.

In addition, I reviewed the [configs/image_median_statistics.json](https://github.com/Project-MONAI/tutorials/blob/main/generation/maisi/configs/image_median_statistics.json) file in the HuggingFace repo and couldn’t find anything related to neck anatomy. Dataset preparation is not an issue on my side, I am generating embeddings  through the MAISI scripts, and masks are 100% verified.

This leaves me confused: do I need to train ControlNet from scratch (and potentially the each model from scratch), or is MAISI simply not designed to handle neck scenarios? Also, what would be a good dataset size to train on? I noticed in discussions that @sara-create had issues even with 300 cases, while @pshavela trained on ~700 cases and reported good results.
A reply on this issue would really help me a lot, much appreciated!

**Reply by Can-Zhao** [reply]
Thank you! I answered it in https://github.com/Project-MONAI/tutorials/discussions/2022
